cmd/kubelet/app/options/options.go

```diff
@@ -292,7 +292,7 @@ func (f *KubeletFlags) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&f.RootDirectory, "root-dir", f.RootDirectory, "Directory path for managing kubelet files (volume mounts,etc).")
 
 	fs.Var(&f.DynamicConfigDir, "dynamic-config-dir", "The Kubelet will use this directory for checkpointing downloaded configurations and tracking configuration health. The Kubelet will create this directory if it does not already exist. The path may be absolute or relative; relative paths start at the Kubelet's current working directory. Providing this flag enables dynamic Kubelet configuration. Presently, you must also enable the DynamicKubeletConfig feature gate to pass this flag.")
-	fs.Var(&f.InitConfigDir, "init-config-dir", "The Kubelet will look in this directory for the init configuration. The path may be absolute or relative; relative paths start at the Kubelet's current working directory. Omit this argument to use the built-in default configuration values. Presently, you must also enable the KubeletConfigFile feature gate to pass this flag.")
+	fs.Var(&f.InitConfigDir, "init-config-dir", "The Kubelet will look in this directory for the init configuration. The path may be absolute or relative; relative paths start at the Kubelet's current working directory. Omit this argument to use the built-in default configuration values. You must also enable the KubeletConfigFile feature gate to pass this flag. Since this feature is in beta, the feature gate is enabled by default.")
```

Comment:
remind me why this points to a directory within which we look for a magic named file? I expected to point to the config file.

Reply:
Also, I'm unsure why we need a flag and a feature gate... nothing else looks at the feature gate, right? I'd expect presence of the flag to be sufficient

Reply:
I was going for isomorphic structure to what you get when you unpack a ConfigMap, so that if you have your on-disk config in this directory, and you want to generate a ConfigMap from it that works for dynamic config, it's a single command away.

The feature gate was probably a little overly cautious on my part.

Reply:
pointing the kubelet at a dir makes sense when the kubelet owns the content of the dir. it makes less sense when you have to have magic knowledge of the filename to populate inside the dir

Reply:
You need to know the correct key names when you put data in a ConfigMap anyway.
Sure you can specify the key names in the command to kubectl if you're generating a ConfigMap from a directory. 
But if you make the key names different in the ConfigMap vs the file names, then relative paths in the new ConfigMap will be wrong, unless you change them to point to your new key names.
So I think it's much simpler to just require the same schema in both places.

Reply:
> You need to know the correct key names when you put data in a ConfigMap anyway.

The user does. The kublet shouldn't. 

> But if you make the key names different in the ConfigMap vs the file names, then relative paths in the new ConfigMap will be wrong, unless you change them to point to your new key names.

I'm not advocating mismatched names, I'm advocating being explicit in telling the kubelet what config file to use, and what key of the configmap to use, rather than having magic knowledge inside the kubelet that decides that.

Reply:
>The kubelet should be told the correct key name.

Having a flag to specify which ConfigMap key the payload lives in feels like having a flag to specify the name of another flag. The additional indirection doesn't seem useful to me. What's wrong with standardizing key names?

Standards make the system easier to talk about and understand. Why shouldn't the structure of the directory be considered an API?

Reply:
> The additional indirection doesn't seem useful to me. 

The point is not indirection, it is making implicit and hidden references explicit.

> Why shouldn't the structure of the directory be considered an API?

That leads to more implicit behavior. For example:
* If a file named `foo` exists in the dir, optionally enable feature X, otherwise disable feature X
* If a file named `foo` exists in the dir, optionally configure feature X, otherwise use default config options

This leads to undesirable failure modes, where you thought you were configuring a feature, but a typo or error means it got silently disabled or misconfigured (for example, the default `--kubeconfig` arg to the kubelet, meaning we couldn't treat failure to load `--kubeconfig` as a real error, leading to `--require-kubeconfig` as a second arg people had to know to specify)

Reply:
>The point is not indirection, it is making implicit and hidden references explicit.

If the ConfigMap key to use is specified on the Kubelet command line, it's still hidden and implicit from the point of view of a remote ConfigMap, and this makes portability of config more difficult.

I want to tell customers "just put it in the `kubelet.config.k8s.io` key."
I don't want to tell customers "go find your Kubelet command line, see what key the Kubelet is configured to look for in a ConfigMap for its configuration, and then use that."

> That leads to more implicit behavior. For example:

If a new binary can recognize a new key, but the key isn't provided, it should just use a default rather than refusing to run because it doesn't see the new key. This kind of implicit behavior is key to decoupling binary updates from configuration changes.

We should have a consistent behavior for missing keys/files, so people know what to expect. Perhaps:
- If the key (or file) doesn't exist, use the defaults.
- Default to empty if you want to disable on omission. The problem with `--kubeconfig` was that it had a non-empty default, thus omission could originally be used as an off switch.

Then at least we have consistent behavior across key names.

It doesn't solve the issue with silent failures resulting from typos in key names. Happy to have the Kubelet log warnings along the lines of "noticed key with name x, which I don't recognize."

Reply:
> If the ConfigMap key to use is specified on the Kubelet command line, it's still hidden and implicit from the point of view of a remote ConfigMap, and this makes portability of config more difficult.

For the configmap key, it's not on the command line, it would be in the `configSource.configMapRef`. The reference would match the thing it is referring to in visibility:
* init config file local to the kubelet, specified by flag
* configmap key in the API, referenced by the Node object's configSource

If we want to pin the name, making it explicit still seems like a good idea (`--init-config-file=foo/bar/kubelet`, and validate that a file named `kubelet` is referenced). It makes it very clear what the config file is, lets most of the stack not need to know about magic pinned filenames, and if you ever want to unpin the name, means all existing consumers have explicit references that continue to be valid.

Reply:
Ok, I can live with it if we lift the mapping into the `configSource.configMapRef`. I was wary of putting more than necessary on the command line, but I see that's not what you were suggesting. Points about being able to unpin names make sense.

I don't have time to change `configSource.configMapRef` for 1.9, but dynamic config is still alpha anyway, so I don't think that should block loading from a file for 1.9.

If I change this PR to take an explicit filename and validate that it exists, is that sufficient for 1.9?